pkg/kube/client.go

```diff
@@ -116,50 +150,101 @@ type Client interface {
 	clientFactory util.Factory
 	restClient    *rest.RESTClient
-	config        *rest.Config
-	extSet        *kubeExtClient.Clientset
 	revision      string
+
+	config *rest.Config
+
+	extSet       *kubeExtClient.Clientset
+	kubeInformer informers.SharedInformerFactory
+
+	dynamic         dynamic.Interface
+	dynamicInformer dynamicinformer.DynamicSharedInformerFactory
+
+	metadata         metadata.Interface
+	metadataInformer metadatainformer.SharedInformerFactory
 }
 
-// NewClient creates a Kubernetes client from the given factory. The "revision" parameter
-// controls the behavior of GetIstioPods, by selecting a specific revision of the control plane.
-func NewClient(clientFactory util.Factory, revision string) (Client, error) {
-	restConfig, err := clientFactory.ToRESTConfig()
+// newClientInternal creates a Kubernetes client from the given factory.
+func newClientInternal(clientFactory util.Factory, revision string) (*client, error) {
+	var c client
+	var err error
+
+	c.revision = revision
+
+	c.restClient, err = clientFactory.RESTClient()
+	if err != nil {
+		return nil, err
+	}
+
+	c.config, err = clientFactory.ToRESTConfig()
 	if err != nil {
 		return nil, err
 	}
-	restClient, err := clientFactory.RESTClient()
+
+	c.Interface, err = kubernetes.NewForConfig(c.config)
 	if err != nil {
 		return nil, err
 	}
-	clientSet, err := kubernetes.NewForConfig(restConfig)
+	c.kubeInformer = informers.NewSharedInformerFactory(c.Interface, resyncInterval)
+
+	c.metadata, err = metadata.NewForConfig(c.config)
+	if err != nil {
+		return nil, err
+	}
+	c.metadataInformer = metadatainformer.NewSharedInformerFactory(c.metadata, resyncInterval)
+
+	c.dynamic, err = dynamic.NewForConfig(c.config)
 	if err != nil {
 		return nil, err
 	}
-	extSet, err := kubeExtClient.NewForConfig(restConfig)
+	c.dynamicInformer = dynamicinformer.NewDynamicSharedInformerFactory(c.dynamic, resyncInterval)
+
+	c.extSet, err = kubeExtClient.NewForConfig(c.config)
 	if err != nil {
 		return nil, err
 	}
-	return &client{
-		clientFactory: clientFactory,
-		Clientset:     clientSet,
-		restClient:    restClient,
-		config:        restConfig,
-		extSet:        extSet,
-		revision:      revision,
-	}, nil
+
+	return &c, nil
 }
 
-// NewClient creates a Kubernetes client from the given ClientConfig. The "revision" parameter
+// NewExtendedClient creates a Kubernetes client from the given ClientConfig. The "revision" parameter
 // controls the behavior of GetIstioPods, by selecting a specific revision of the control plane.
-func NewClientForConfig(clientConfig clientcmd.ClientConfig, revision string) (Client, error) {
-	return NewClient(newClientFactory(clientConfig), revision)
+func NewExtendedClient(clientConfig clientcmd.ClientConfig, revision string) (ExtendedClient, error) {
+	return newClientInternal(newClientFactory(clientConfig), revision)
+}
+
+// NewClient creates a Kubernetes client from the given rest config.
+func NewClient(config *rest.Config) (Client, error) {
```

Comment:
Any reason why we're changing this to `rest.Config` from `util.Factory`? The callers can convert easily enough and `util.Factory` is ultimately what is needed.

I think both `NewClient` and `NewExtendedClient` should just take `util.Factory` to be consistent. We can also make `istioKube.newClientFactory` public if that's useful.

Reply:
The good news is we will only call this in one place now, so if the caller has to do more work its not a big deal - I'll change this to factory